modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  on communication.
+*/
+
+/*
+  Below are segment statuses, which is a way to make visible to outsiders the
+  current ongoing operation.
+*/
+private param STATUS_UNLOCKED : uint = 0;
+private param STATUS_ADD : uint = 1;
+private param STATUS_REMOVE : uint = 2;
+private param STATUS_LOOKUP : uint = 3;
+private param STATUS_BALANCE : uint = 4;
+
+/*
+  Below are statuses specific to the work stealing algorithm.
+*/
+private param WS_INITIALIZED = -1;
+private param WS_FINISHED_WITH_NO_WORK = 0;
+private param WS_FINISHED_WITH_WORK = 1;
+
+/*
+  The phases for operations.
+*/
+private param ADD_BEST_CASE = 0;
+private param ADD_AVERAGE_CASE = 1;
+private param REMOVE_BEST_CASE = 2;
+private param REMOVE_AVERAGE_CASE = 3;
+private param REMOVE_WORST_CASE = 4;
+
+/*
+  The initial amount of elements in an unroll block. Each successive unroll block
+  is double the size of it's predecessor, allowing for better locality for when
+  there are larger numbers of elements. The better the locality, the better raw
+  performance and easier it is to redistribute work.
+*/
+config param distributedBagInitialBlockSize = 1024;
+/*
+  To prevent stealing too many elements (horizontally) from another node's segment
+  (hence creating an artifical load imbalance), if the other node's segment has
+  less than a certain threshold (see `distributedBagWorkStealingMemCap`) but above
+  another threshold (see `distributedBagWorkStealingMinElems`), we steal a percentage of their
+  elements, leaving them with majority of their elements. This way, the amount the
+  other segment loses is proportional to how much it owns, ensuring a balance.
+*/
+config param distributedBagWorkStealingRatio = 0.25;
+/*
+  The maximum amount of work to steal from a horizontal node's segment. This
+  should be set to a value, in megabytes, that determines the maximum amount of
+  data that should be sent in bulk at once. The maximum number of elements is
+  determined by: (distributedBagWorkStealingMemCap * 1024 * 1024) / sizeof(eltType).
+  For example, if we are storing 8-byte integers and have a 1MB limit, we would
+  have a maximum of 125,000 elements stolen at once.
+*/
+config param distributedBagWorkStealingMemCap : real = 1.0;
+/*
+  The minimum number of elements a horizontal segment must have to become eligible
+  to be stolen from. This may be useful if some segments produce less elements than
+  others and should not be stolen from.
+*/
+config param distributedBagWorkStealingMinElems = 1;
+
+/*
+  A parallel-safe distributed multiset implementation that scales in terms of
+  nodes, processors per node (PPN), and workload; The more PPN, the more segments
+  we allocate to increase raw parallelism, and the larger the workload the better
+  locality (see `distributedBagInitialBlockSize`). This data structure is unordered and employs
+  its own work-stealing algorithm, and provides a means to obtain a privatized instance of
+  the data structure for maximized performance.

```

Comment:
This paragraph is a pretty good introduction. For full-on user documentation you'd also need to include an example of using it.